rails/spec/controllers/admin/project_links_controller_spec.rb

```diff
@@ -0,0 +1,325 @@
+
+    describe 'Show' do
+      describe 'link 1 (user IS a project admin)' do
+        it 'lets them see it' do
+          get :show, id:@link_1.id
+          expect(assigns[:project_link]).to eq(@link_1)
+          expect(response).to have_http_status(:ok)
+        end
+      end
+      describe 'link 2 (user is NOT a project admin)' do
+        it 'wont let them see it' do
+          get :show, id:@link_2.id
+          # Redirect, and show error when not allowed:
+          expect(response).to have_http_status(:redirect)
+          expect(request.flash[:alert]).to match(RegexForAuthFailShow)
+        end
+      end
+    end
+
+    describe 'New' do
+      it 'it should show the New form' do
+        expect(get :new).to render_template('new')
+        # Redirect, and show error when not allowed:
+      end
+      it 'should display only the projects for which the user is an admin' do
+        get :new
+        expect(assigns(:projects)).to include(project_1)
+        expect(assigns(:projects)).not_to include(project_2)
+        expect(assigns(:projects)).not_to include(project_3)
+      end
+      it 'should return an OK http status' do
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    describe 'Edit' do
+      before(:each) do
+        allow(Admin::ProjectLink).to receive(:find).and_return(@link_1)
+      end
+      it 'it should show the Edit form' do
+        expect(get :edit).to render_template('edit')
+        # Redirect, and show error when not allowed:
+      end
+      it 'should display only the projects for which the user is an admin' do
+        get :edit
+        expect(assigns(:projects)).to include(project_1)
+        expect(assigns(:projects)).not_to include(project_2)
+        expect(assigns(:projects)).not_to include(project_3)
+      end
+      it 'should return an OK http status' do
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    describe 'Create' do
+      let(:params) do
+        {
+          admin_project_link: {
+            project_id: project_id,
+            name: 'name',
+            href: 'http://foo.com/',
+            link_id: 'foo'
+          }
+        }
+      end
+
+      describe 'when creating a link for project 1 which they ARE admin for' do
+        let(:project_id) { project_1.id }
+        it 'it SHOULD let them' do
+          post :create, params
+          link = assigns(:project_link)
+          expect(assigns(:project_link)).to be_valid
+          expect(response).to redirect_to(admin_project_link_path(link))
+        end
+      end
+
+      describe 'when creating a link for project 2 (NOT their project)' do
+        let(:project_id) { project_2.id }
+        it 'it should NOT let them' do
+          post :create, params
+          # Redirect, and show error when not allowed:
+          expect(response).to have_http_status(:redirect)
+          expect(request.flash[:alert]).to match(RegexForAuthFailNew)
+        end
+      end
+    end
```

Comment:
There is no test of update here. It should test that a project admin can't change the project to a different project than their own. Which I think will currently fail.